src/nunit.analyzers/ValueSourceUsage/ValueSourceUsageAnalyzer.cs

```diff
@@ -0,0 +1,171 @@
+            defaultSeverity: DiagnosticSeverity.Error,
+            description: ValueSourceUsageConstants.SourceDoesNotReturnIEnumerableDescription);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(
+            considerNameOfDescriptor,
+            missingSourceDescriptor,
+            sourceNotStaticDescriptor,
+            methodExpectParameters,
+            sourceDoesNotReturnIEnumerable);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+
+            context.RegisterSyntaxNodeAction(x => AnalyzeAttribute(x), SyntaxKind.Attribute);
+        }
+
+        private static void AnalyzeAttribute(SyntaxNodeAnalysisContext context)
+        {
+            var attributeInfo = SourceHelpers.GetSourceAttributeInformation(
+                context,
+                NunitFrameworkConstants.FullNameOfTypeValueSourceAttribute,
+                NunitFrameworkConstants.NameOfValueSourceAttribute);
+
+            if (attributeInfo == null)
+            {
+                return;
+            }
+
+            var stringConstant = attributeInfo.SourceName;
+            var syntaxNode = attributeInfo.SyntaxNode;
+
+            if (syntaxNode == null || stringConstant == null)
+            {
+                return;
+            }
+
+            var symbol = SourceHelpers.GetMember(context, attributeInfo);
+            if (symbol is null)
+            {
+                context.ReportDiagnostic(Diagnostic.Create(
+                    missingSourceDescriptor,
+                    syntaxNode.GetLocation(),
+                    stringConstant));
+            }
+            else
+            {
+                var sourceIsAccessible = context.SemanticModel.IsAccessible(
+                    syntaxNode.GetLocation().SourceSpan.Start,
+                    symbol);
+
+                if (attributeInfo.IsStringLiteral && sourceIsAccessible)
+                {
+                    var nameOfClassTarget = attributeInfo.SourceType.ToMinimalDisplayString(
+                        context.SemanticModel,
+                        syntaxNode.GetLocation().SourceSpan.Start);
+
+                    var nameOfTarget = attributeInfo.SourceType == context.ContainingSymbol.ContainingType
+                        ? stringConstant
+                        : $"{nameOfClassTarget}.{stringConstant}";
+
+                    var properties = new Dictionary<string, string>
+                    {
+                        { SourceCommonConstants.PropertyKeyNameOfTarget, nameOfTarget }
+                    };
+
+                    context.ReportDiagnostic(Diagnostic.Create(
+                        considerNameOfDescriptor,
+                        syntaxNode.GetLocation(),
+                        properties.ToImmutableDictionary(),
+                        nameOfTarget,
+                        stringConstant));
+                }
```

Comment:
Could you extract that to separate method? E.g. `ReportToUseNameOfIfApplicable`.

Quite a lot happens in this method, a bit hard to keep track

Reply:
Sounds like a good suggestion, then we can also use it for the `TestCaseSourceUsesStringAnalyzer`